### ISSUE TITLE
Provide an easier way to defer starting the app with FlutterActivity

### ISSUE THREAD

**Issue by collinjackson** [flutter/flutter#11239]
If you're extending `FlutterActivity` you might want to do some processing in Java before starting up the Flutter app. Right now, `FlutterActivity` assumes you want to start the app immediately. You could have a different `Activity` that transitions to the `FlutterActivity`, but that could involve some extra work, and it would be nice to be able to do it in `FlutterActivity`.

Android requires that `onCreate` calls `super.onCreate()`. If `FlutterActivity` had way to call `super.onCreate()` without `eventDelegate.onCreate(savedInstanceState);` (perhaps by exposing the delegate to subclasses or by notifying the `eventDelegate` via protected methods that could be overridden), that would allow subclasses of `FlutterActivity` to delay startup until they are ready for the Flutter app to start.

I believe the current recommended solution is copy-pasting the entirety of `FlutterActivity`. This works, but it is adding extra boilerplate that might get out of date.

This issue is possibly related to #11237 (checking for Google Play Services on startup). It was based on a conversation with @lukef. It sounds like Posse has worked around it already by making their own Activity, so this is just feedback for the next developer who comes along. @lukef, I'm not sure if I correctly accurately the issue you described, so feel free to chime in with more thoughts at your convenience.

**Comment by matthew-carroll** [comment]
Could the desired effect be achieved with a `FlutterFragment`?  The specific use-case isn't described in this ticket, but a `FlutterFragment` could be created at any later point the developer desired.

If a Fragment won't cut it then I think we need to log more details about what is desired here.

CC @Hixie @mehmetf

**Comment by mehmetf** [comment]
Thinking out loud here.

One way would be to provide all the functionality around a single native object; preferably the lowest level one that add2app developers interact with. That would mean wherever Flutter is used, the *right things happen* at all times. On Android, this construct is a FlutterView but since Views do not really have a lifecycle, it is impossible to provide complete functionality. Fragment might be the right answer. The complication comes in for cases where add2app really wants a view. One example is placing Flutter inside of a RecyclerView. So, in this scenario, how could we handle Flutter functionality that requires hooking into the activity lifecycle?

That brings me to the second way which is already halfway implemented: a delegate pattern. Push all Flutter related functionality to FlutterActivityDelegate and leave it up to the host app to do the necessary wiring. We can then provide adequate instructions on how to do wiring for various add2app scenarios involving Activity, Fragment and View. I would then make FlutterActivity in the templates just a shell; the simplest possible incarnation of an actual activity using the delegate.

**Comment by lukef** [comment]
The issue mentioned wasn't due to a specific architectural issue with an Activity vs Fragment vs View but more an issue surrounding the fact that in the `Application` class we wanted to defer the "boot" of flutter until various other checks had been performed (namely a check for GMS and one other check, if i recall correctly).

The problem with Flutter is that the start up will happen async and disconnected to the native start up. Things get out of sync very easily. I know I've had this discussion a few times and there isn't a super clear path as I understand it. The request, in this case, was to just defer starting anything flutter related until a time of our choosing.

Not sure if i'm interpreting "add2app" correctly, but this is also something relevant to "pure" flutter apps also. In the case of GMS, to use that as an example, you have to start flutter with a "loading screen" which delays the overall check until flutter starts which isn't totally ideal but workable. If your app depends on GMS and it's not available then it shouldn't even start realistically.

The reality of flutter is that is HAS to work with the native ecosystem / lifecycles / best practices as well. If it doesn't then it'll create inefficiencies.

**Comment by lukef** [comment]
As a side, there are also specific use cases where you may need to communicate with flutter from the native side and flutter has not yet started (deep linking, startup checks, etc). I know I talked to @xster a whole bunch about this and whether it'd be possible to cache messages from native to flutter. I only mention this because it's most likely connected.

**Comment by matthew-carroll** [comment]
@lukef Can you clarify whether this issue is referring to the Flutter initialization within the Application class vs the FlutterView and FlutterNativeView initialization that occurs within an Activity?

The original write-up mentions Activity exclusively, but your recent comment mentions Application...

**Comment by matthew-carroll** [comment]
I'm closing this ticket because it's unclear which initialization it refers to.  If the issue still exists after the Android embedding refactor is merged in then please file a new ticket against that API.
